### PR TITLE
UI: User invite language

### LINKED ISSUE #2829: Email invitations and landing page (language settings)
@minotogna 

Currently, it seems that the email invitation is only available in English. Please double-check if we have email invitations in other UN languages(we could select languages when sending the invitation). In case we do not, please send an email to the corresponding colleagues. Many thanks!

### PR DESCRIPTION
**Pull request by sorja**
[screen-recorder-fri-aug-25-2023-13-46-10.webm](https://github.com/openforis/fra-platform/assets/5508251/ef1ed005-69a8-4dc4-8a19-8252c82d5744)

### CODE REVIEW

**Comment by minotogna** [comment]
I think that what the issue meant is 
1: invitation email is received in the language of the inviter 
2: the invitation page must be opened in the language of the invitation . Plus the possibility to switch language

File: src/client/components/InviteUserForm/InviteUserForm.tsx [hunk 1]
```diff
@@ -21,26 +23,38 @@ const validateEmail = (email: string) => {
...
 
-  const onUserInvite = () => {
+  const langs = useMemo(() => {
+    // default to English as first option
+    return LanguageCodes.sort((a, b) => {
+      if (a === Lang.en) return -1
+      if (b === Lang.en) return 1
+      return 0
+    })
+  }, [])
```
Comment (minotogna) [inline src/client/components/InviteUserForm/InviteUserForm.tsx]: lagna must be ordered in the same way as language selector. use the already available array

File: src/client/components/InviteUserForm/InviteUserForm.tsx [hunk 2]
```diff
@@ -55,72 +69,104 @@ const InviteUserForm: React.FC = () => {
...
-          toaster.info(i18n.t('userManagement.userAdded', { email: userToInvite.email }))
+          setUserToInvite({ name: '', email: '', role: '', lang: Lang.en })
+          toaster.info(t('userManagement.userAdded', { email: userToInvite.email }))
         })
         .catch(() => {
           // Error handled by server
         })
   }
 
+  const onUserInvite = useCallback(_onUserInvite, [userToInvite, countryIso, assessment, cycle, t, toaster, dispatch])
```
Comment (minotogna) [inline src/client/components/InviteUserForm/InviteUserForm.tsx]: export this as useOnInvite

File: src/client/components/InviteUserForm/InviteUserForm.tsx [hunk 3]
```diff
@@ -55,72 +69,104 @@ const InviteUserForm: React.FC = () => {
...
+          role: userToInvite.role as RoleName,
           email: userToInvite.email,
+          lang: userToInvite.lang,
         })
       )
         .unwrap()
         .then(() => {
-          setUserToInvite({ name: '', email: '', role: '' })
-          toaster.info(i18n.t('userManagement.userAdded', { email: userToInvite.email }))
+          setUserToInvite({ name: '', email: '', role: '', lang: Lang.en })
```
Comment (minotogna) [inline src/client/components/InviteUserForm/InviteUserForm.tsx]: default lang should be the inviter user language

File: src/client/components/InviteUserForm/InviteUserForm.tsx [hunk 4]
```diff
@@ -21,26 +23,38 @@ const validateEmail = (email: string) => {
...
   email: string
+  lang: Lang
 }
 
 const InviteUserForm: React.FC = () => {
   const [userToInvite, setUserToInvite] = useState<UserToInvite>({
     name: '',
     email: '',
+    lang: Lang.en,
   })
```
Comment (minotogna) [inline src/client/components/InviteUserForm/InviteUserForm.tsx]: default lang should be the inviter user language. and use only one initial state (see line 79)

File: src/client/components/UserList/UserListHeader/UserListHeader.tsx [hunk 5]
```diff
@@ -31,6 +40,11 @@ const UserListHeader: React.FC<{ isAdmin: boolean; readOnly: boolean }> = ({ isA
         {!readOnly && (
           <th className="user-list__header-cell user-list__edit-column">
             <UserListButtonExport isAdmin={isAdmin} />
+            {Users.getRolesAllowedToEdit({ user, countryIso, cycle }).length > 0 && (
+              <Link to="invite" className="btn-s btn-primary">
```
Comment (minotogna) [inline src/client/components/UserList/UserListHeader/UserListHeader.tsx]: do not hardcode paths, access routes  always through Routes

File: src/client/components/UserList/UserListHeader/UserListHeader.tsx [hunk 6]
```diff
@@ -1,25 +1,34 @@
...
 
+import { useCycle } from 'client/store/assessment'
 import { useFilteredRoleNames } from 'client/store/ui/userManagement'
+import { useUser } from 'client/store/user'
+import { useCountryRouteParams } from 'client/hooks/useRouteParams'
+import Icon from 'client/components/Icon'
 
 import UserListButtonExport from '../UserListButtonExport'
 
 const UserListHeader: React.FC<{ isAdmin: boolean; readOnly: boolean }> = ({ isAdmin, readOnly }) => {
```
Comment (minotogna) [inline src/client/components/UserList/UserListHeader/UserListHeader.tsx]: extract Props type and isAdmin can be derived from useUser and always use props param, not spreaded

File: src/client/pages/CountryHome/FraHome/Collaborators/Collaborators.tsx [hunk 7]
```diff
@@ -1,25 +1,18 @@
...
-  const countryIso = useCountryIso()
-  const cycle = useCycle()
 
   return (
-    <>
-      {Users.getRolesAllowedToEdit({ user, countryIso, cycle }).length > 0 && <InviteUserForm />}
-      <UserList users={users} />
-    </>
+    <Routes>
+      <Route path="invite" element={<InviteUserForm />} />
```
Comment (minotogna) [inline src/client/pages/CountryHome/FraHome/Collaborators/Collaborators.tsx]: do not hardcode Routes

File: src/client/pages/CountryHome/FraHome/FraHome.tsx [hunk 8]
```diff
@@ -54,7 +54,7 @@ const FraHome: React.FC = () => {
 
       <Routes>
         {sections.map(({ name, component }) => (
-          <Route key={name} path={name} element={React.createElement(component, {})} />
+          <Route key={name} path={`${name}/*`} element={React.createElement(component, {})} />
```
Comment (minotogna) [inline src/client/pages/CountryHome/FraHome/FraHome.tsx:57]: I really don't like this *. just to mention, we need another refactor later on

File: src/server/api/user/invite.ts [hunk 9]
```diff
@@ -12,11 +13,12 @@ export const invite = async (
     email: string
     name: string
     role: RoleName
+    lang: Lang
```
Comment (minotogna) [inline src/server/api/user/invite.ts]: extract type InviteUserRequest (in this file). it's more readable

File: src/server/controller/user/invite.ts [hunk 10]
```diff
@@ -32,7 +32,7 @@ export const invite = async (
     // If neither of above, create new user
     if (!userToInvite)
       userToInvite = await UserRepository.create(
-        { user: { email, props: { name: name ?? '', lang: user.props.lang ?? Lang.en } } },
+        { user: { email, props: { name: name ?? '', lang: lang ?? Lang.en } } },
```
Comment (minotogna) [inline src/server/controller/user/invite.ts]: isn't lang required ?

File: src/server/repository/public/user/create.ts [hunk 11]
```diff
@@ -14,7 +15,7 @@ export const create = async (
     user: { props: userProperties, email },
   } = props
 
-  if (!userProperties.lang) userProperties.lang = 'en'
+  if (!userProperties.lang) userProperties.lang = Lang.en
```
Comment (minotogna) [inline src/server/repository/public/user/create.ts]: why should this be undefined ?

**Comment by minotogna** [comment]
invite user backgorund page, better to be white

File: src/client/components/InviteUserForm/InviteUserForm.tsx (hunk 1 above)
Comment (sorja) [reply src/client/components/InviteUserForm/InviteUserForm.tsx]: defaults to english though?

**Comment by minotogna** [comment]
input field should have a horizontal padding, like other form inputs. don't we have a generic rule ?

**Comment by minotogna** [comment]
the submit button, better to say. `+ Invite`

**Comment by minotogna** [comment]
> the submit button, better to say. `+ Invite`

Or just 'Submit' ?

**Comment by minotogna** [comment]
@sorja after submitting, it might be useful to navigate back to manage collaborators page.

File: src/meta/routes/routes.ts [hunk 12]
```diff
@@ -59,6 +60,7 @@ export const Routes = {
   CountryDataDownload,
   CountryHome,
   CountryHomeSection,
+  CountryHomeInviteSection,
```
Comment (minotogna) [inline src/meta/routes/routes.ts]: maybe I would call this CountryHomeSectionInvite

File: src/client/components/UserList/InviteUserLink/InviteUserLink.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+  const { className } = props
+  const { t } = useTranslation()
+  const user = useUser()
+  const { countryIso } = useCountryRouteParams()
+  const cycle = useCycle()
+
+  if (!Users.getRolesAllowedToEdit({ user, countryIso, cycle }).length) return null
+
+  return (
+    <Link to="invite" className={classNames('btn-s', 'btn-primary', className)}>
```
Comment (minotogna) [inline src/client/components/UserList/InviteUserLink/InviteUserLink.tsx]: better to use Routes, not hardcoded path

**Comment by sorja** [comment]
firefox_a2cNvOY4Pt

@minotogna  Added redirect

**Review by minotogna — APPROVED** [review]

**Pull request merged**